[#1581] Click on resource tag adds to filter

andrewjprice opened the pull request:
<!--[
  Thank you for contributing! Please use this pull request (PR) template.

  Need help? Post in the #dev channel on Slack
  Use the "wip" label if this PR is not ready for review

  Check out our Pull Request Practices guide if you haven't already: https://github.com/ifmeorg/ifme/wiki/Pull-Request-Practices
  Join our organization if you haven't already: https://github.com/ifmeorg/ifme/wiki/Join-Our-Slack
  We encourage everyone to add themselves to our Contribute page: https://github.com/ifmeorg/ifme/wiki/Contributor-Blurb
]-->
# Description

Clicking any tag on the [resource page](https://www.if-me.org/resources) will add it to the current filter.

## Corresponding Issue

[#1581]

# Screenshots

![resource-tag-to-filter](https://user-images.githubusercontent.com/22504617/66625301-86cb1500-eba8-11e9-833e-28d85d0415b8.gif)

# Test Coverage

✅ Passing
---

Reviewing this pull request? Check out our [Code Review Practices](https://github.com/ifmeorg/ifme/wiki/Code-Review-Practices) guide if you haven't already!

julianguyen commented on client/app/components/Tag/Tag.scss:
```diff
@@ -13,6 +13,7 @@
   border: $size-1 solid $mulberry;
   mix-blend-mode: normal;
   color: $mulberry;
+  cursor: pointer;
```
Nice!

julianguyen commented on client/app/components/Tag/index.jsx:
```diff
@@ -7,14 +7,15 @@ type Props = {
     normal ? css.normal : ''
   } ${secondary ? css.secondary : ''}`;
-  return <div className={labelClassNames}>{label}</div>;
+  return <div onClick={() => onClick(label)} className={labelClassNames}>{label}</div>;
```
So if this function doesn't exist, we should execute nothing. You'll want to do something like this: `onClick={onClick ? () => onClick(label) : null}`. I think this is what is responsible for the Flow error that's happening in the CircleCI build.

julianguyen commented on client/app/components/Resource/index.jsx:
```diff
@@ -11,14 +11,15 @@ type Props = {
       <div className="tags">
         {tags.map((tag) => (
-          <Tag normal label={tag} key={Utils.randomString()} />
+          <Tag normal label={tag} key={Utils.randomString()} onClick={ (tagLabel) => { updateTagFilter(tagLabel) } } />
```
You can re-write this as `onClick={updateTagFilter ? (tagLabel) => updateTagFilter(tagLabel) : null}`. I think this is what is responsible for the Flow error that's happening in the CircleCI build.

julianguyen commented on client/app/widgets/Resources/index.jsx:
```diff
@@ -110,6 +110,13 @@ export class Resources extends React.Component<Props, State> {
+  updateTagFilter = (tagLabel: String) => {
+    this.setState((prevState: State) => {
+      const updatedBoxes = prevState.checkboxes.map((checkbox) => checkbox.label === tagLabel ? {...checkbox, checked: true } : checkbox)
+      return { checkboxes: updatedBoxes };
```
We should write an integration test for this! You'll want to test that clicking tags changes the checkboxes.

julianguyen commented:
You can run `yarn lint` in your locally environment to reproduce what's happening in the Circle CI build if you haven't already done that.

andrewjprice commented:
@julianguyen I refactored some pieces and added a test to ensure clicking a tag adds it to the filter. Please look it over and let me know if this is what is expected.

There are currently 3 tests failing which I'm having trouble resolving. I think 2 are failing due to the method `static getDerivedStateFromProps`. I'm unsure of the third. Help is appreciated 🙂 .

julianguyen commented on client/app/components/Input/InputTag.jsx:
```diff
@@ -29,6 +29,13 @@ export class InputTag extends React.Component<Props, State> {
     this.state = { checkboxes: props.checkboxes, autoHighlight: false };
   }
 
+  static getDerivedStateFromProps(nextProps: Props, prevState: State) {
```
So this works, but triggers extra, unwanted re-renders. We can remove this and add `key={Utils.randomString()}` to `<InputTag />` in `/client/app/widgets/Resources/index.jsx`. This trigger a re-render when the `state.checkboxes` changes (since it's passed into InputTag).
